### PR TITLE
Tezos support

### PR DESCRIPTION
**Pull request by catenocrypt**

### CODE REVIEW

File: base58.go [hunk 1]
```diff
@@ -45,16 +51,20 @@ func Base58Decode(b, alphabet string) ([]byte, error) {
...
-	checksum := doubleSha256(val[0 : len(val)-4])
-	expected := val[len(val)-4:]
+	checksum := doubleSha256(val[: len(val) - 4])
+	expected := val[len(val) - 4:]
 	if !bytes.Equal(checksum[0:4], expected) {
 		return nil, fmt.Errorf("Bad Base58 checksum: %v expected %v", checksum, expected)
 	}
-	return val, nil
+	// strip checksum
+	return val[:len(val) - 4], nil
```
Comment (hewigovens) [inline base58.go]: can you revert this change? let caller to remove the checksum, we may add hasher to this func to support checksum other than `doubleSha256`

File: tezos.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+func init() {
+	toBytesMap[slip44.TEZOS] = TezosDecodeToBytes
+	toStringMap[slip44.TEZOS] = TezosEncodeToString
+}
+
+// TezosDecodeToBytes converts the input string to a byte array
+func TezosDecodeToBytes(input string) ([]byte, error) {
+	decoded, err := Base58Decode(input, Base58DefaultAlphabet)
```
Comment (hewigovens) [inline tezos.go]: for default alphabet, let's use `Decode` in `btcutil/base58`

File: tezos.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	if err != nil {
+		return nil, err
+	}
+	if len(decoded) != AddressLength {
+		return nil, errors.New("Invalid length")
+	}
+	// no prefix check
```
Comment (hewigovens) [inline tezos.go]: I think we need to check prefix

File: tezos.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	return decoded, nil
+}
+
+// TezosEncodeToString converts the input byte array to a string representation of the Cosmos address.
+func TezosEncodeToString(bytes []byte) (string, error) {
+	if len(bytes) != AddressLength {
+		return "", errors.New("Invalid decoded address length")
+	}
+	encoded := Base58Encode(bytes, Base58DefaultAlphabet)
```
Comment (hewigovens) [inline tezos.go]: try `Encode` in `btcutil/base58`

File: tezos_test.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+			output: "06a1a1a7f2ff4762f8f26aac80221d73be67709dea1d14",
+		},
+		{
+			name:   "Normal3",
+			input:  "tz3VEZ4k6a4Wx42iyev6i2aVAptTRLEAivNN",
+			output: "06a1a461af383a78291ace2dea59d3da6c9a8b1cdb1b96",
+		},
+		{
+			name:  "Invalid prefix, valid checksum",
+			input: "NmH7tmeJUmHcncBDvpr7aJNEBk7rp5zYsB1qt",
```
Comment (hewigovens) [inline tezos_test.go]: valid tezos address? should be error?

File: tezos_test.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+			err:   errors.New("Bad Base58 checksum: [165 203 237 246 198 234 235 145 5 193 229 5 152 68 252 22 56 193 215 76 132 138 239 45 248 112 249 192 35 164 195 55] expected [57 103 234 47]"),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := TezosDecodeToBytes(tt.input)
+			if tt.err != nil {
+				goterror := "(no error)"
+				if err != nil { goterror = err.Error() }
+				if goterror != tt.err.Error() {
```
Comment (hewigovens) [inline tezos_test.go]: you can just check error prefix, see `xrp_test.go`

**Comment by hewigovens** [comment]
don't forget to run gofmt and golint

File: base58.go (hunk 1 above)
Comment (catenocrypt) [reply base58.go]: Well, I disagree.  Either just to base58 decode, or do checksum change and remove the checksum.  I refactor in Base58 only and Base58 with checksum.

File: tezos.go (hunk 2 above)
Comment (catenocrypt) [reply tezos.go]: For Tezos we need default alphabet, and checksum.  The btcsuite/btcutil uses default alphabet, and has a checksum version, but that also uses one byte version.  So I don't think it's a good idea to use btcsuite/btcutil for Tezos.

File: tezos.go (hunk 3 above)
Comment (catenocrypt) [reply tezos.go]: OK, accepted

File: tezos_test.go (hunk 6 above)
Comment (catenocrypt) [reply tezos_test.go]: Already done

**Comment by catenocrypt** [comment]
azp /run

File: base58.go [hunk 7]
```diff
@@ -1,18 +1,22 @@
...
-	"crypto/sha256"
 	"fmt"
 	"math/big"
 	"strings"
 )
 
+const (
+	// Default alphabet, used e.g. by Bitcoin
+	Base58DefaultAlphabet = 
+	   "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
```
Comment (hewigovens) [inline base58.go]: since we are here, let's remove reference to `btcutil/base58`? add `https://github.com/btcsuite/btcutil/blob/master/base58/alphabet.go` and xrp alphabet for better performance

File: tezos.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
+package coincodec
+
+import (
+	"errors"
+	"bytes"
+	"github.com/wealdtech/go-slip44"
+)
+
+const (
+	AddressLength = 23
```
Comment (hewigovens) [inline tezos.go]: this would expose a general `AddressLength` outside this file, let's add `tezos` prefix to avoid that

Comment (catenocrypt) [reply tezos.go]: done
